fuel-tx/src/tests/valid_cases/transaction/upgrade.rs

```diff
@@ -0,0 +1,729 @@
+        Err(ValidityError::TransactionUpgradeNoPrivilegedAddress),
+        result
+    );
+}
+
+#[test]
+fn check__errors_when_consensus_parameters_invalid_witness_index() {
+    let block_height = 1000.into();
+    let mut tx = valid_upgrade_transaction().finalize();
+
+    // Given
+    *tx.upgrade_purpose_mut() = UpgradePurpose::ConsensusParameters {
+        witness_index: u16::MAX,
+        checksum: Default::default(),
+    };
+
+    // When
+    let result = tx.check(block_height, &test_params());
+
+    // Then
+    assert_eq!(
+        Err(ValidityError::InputWitnessIndexBounds {
+            index: u16::MAX as usize
+        }),
+        result
+    );
+}
+
+#[test]
+fn check__errors_when_consensus_parameters_invalid_checksum() {
+    let block_height = 1000.into();
+
+    // Given
+    let mut tx = valid_upgrade_transaction()
+        .add_witness(vec![123; 1024].into())
+        .finalize();
+    *tx.upgrade_purpose_mut() = UpgradePurpose::ConsensusParameters {
+        witness_index: 0,
+        checksum: Default::default(),
+    };
+
+    // When
+    let result = tx.check(block_height, &test_params());
+
+    // Then
+    assert_eq!(
+        Err(ValidityError::TransactionUpgradeConsensusParametersChecksumMismatch),
+        result
+    );
+}
+
+#[test]
+fn check__errors_when_consensus_parameters_unable_decode_consensus_parameters() {
+    let block_height = 1000.into();
+    let serialized_consensus_parameters = vec![123; 1024];
+
+    // Given
+    let mut tx = valid_upgrade_transaction()
+        .add_witness(serialized_consensus_parameters.clone().into())
+        .finalize();
+    *tx.upgrade_purpose_mut() = UpgradePurpose::ConsensusParameters {
+        witness_index: 0,
+        checksum: Hasher::hash(serialized_consensus_parameters.as_slice()),
+    };
+
+    // When
+    let result = tx.check(block_height, &test_params());
+
+    // Then
+    assert_eq!(
+        Err(ValidityError::TransactionUpgradeConsensusParametersDeserialization),
+        result
+    );
+}
+
+#[test]
+fn check__errors_when_consensus_parameters_different_than_calculated_metadata() {
+    let block_height = 1000.into();
+    let serialized_consensus_parameters = postcard::to_allocvec(&test_params()).unwrap();
+
+    // Given
+    // `valid_upgrade_transaction` already returns a transaction with calculated metadata.
+    // Setting a new `UpgradePurpose` below will cause mismatch between the calculated
+    // metadata and the actual metadata.
+    let mut tx = valid_upgrade_transaction()
+        .add_witness(serialized_consensus_parameters.clone().into())
+        .finalize();
+    *tx.upgrade_purpose_mut() = UpgradePurpose::ConsensusParameters {
+        witness_index: 0,
+        checksum: Hasher::hash(serialized_consensus_parameters.as_slice()),
+    };
+
+    // When
+    let result = tx.check(block_height, &test_params());
+
+    // Then
+    assert_eq!(Err(ValidityError::TransactionMetadataMismatch), result);
+}
+
+mod inputs {
```

Comment:
These tests are missing the `check` prefix.

Also, why are these in a separate mod?

Reply:
It is a separate module since all tests below used different input types but testing the same thing, while above we are testing validity rules  from specification.